### PR TITLE
Supports history variables in function contracts

### PR DESCRIPTION
**Pull request by ArenBabikian**
This PR adds support for history variables in `__CPROVER_ensures` contracts. Specifically, users may now access the pre-function call state of a variable within the post-condition. 

Let us consider the following example. Semantically, this example shows a case where the post-condition of a function call ensures that ( the post-call value of `x` ) is greater than ( the pre-call value of `x` + 2 ).
```
void foo(int *x) __CPROVER_assigns(*x)
  __CPROVER_requires(*x > 0)
  __CPROVER_ensures(*x > __CPROVER_old(*x) + 2)
{
  *x = *x + 2;
}
```
In this case, the variable that we are interested in is `x`, which is modified by the call to `foo(int *x)`. The pre-condition contract (`__CPROVER_requires`) may only access the pre-state of the variable `x` (this is done through the inclusion of `*x`). However, the post-condition (`__CPROVER_ensures`) accesses both the post-state of `x` (through `*x`) and the pre-state of `x` (through `__CPROVER_old(*x)`)

**Note:**
We currently only support pointers within the `__CPROVER_old()` construct. In the near future, we plan to add support for symbols and struct members.

---

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: regression/contracts/history-pointer-enforce-03/test.desc
Comment (SaswatPadhi) [inline regression/contracts/history-pointer-enforce-03/test.desc]: What's the `_` for? Is it a typo?

Comment (SaswatPadhi) [reply regression/contracts/history-pointer-enforce-03/test.desc]: For consistency, could you please remove `--trace`? It's not used in other `.desc` files in this PR.

File: regression/contracts/history-pointer-enforce-03/main.c
Comment (SaswatPadhi) [inline regression/contracts/history-pointer-enforce-03/main.c]: Nitpick: Could you please move `__CPROVER_ensures` to next line?

File: regression/contracts/history-pointer-replace-01/test.desc
Comment (SaswatPadhi) [inline regression/contracts/history-pointer-replace-01/test.desc]: "assert this assumption" is confusing. We just assume the post condition, right?

May be just remove that sentence?

Comment (SaswatPadhi) [reply regression/contracts/history-pointer-replace-01/test.desc]: Also, missing newline at the end of the file.

File: regression/contracts/history-pointer-replace-02/test.desc
Comment (SaswatPadhi) [inline regression/contracts/history-pointer-replace-02/test.desc]: "assert this assumption" is confusing. We just assume the post condition, right?

May be just remove that sentence?

Comment (SaswatPadhi) [reply regression/contracts/history-pointer-replace-02/test.desc]: Also, missing newline at the end of the file.

File: regression/contracts/history-pointer-replace-02/main.c
Comment (SaswatPadhi) [inline regression/contracts/history-pointer-replace-02/main.c]: The contract is incorrect here -- should be `__CPROVER_ensures(*x >= __CPROVER_old(*x) + 2)`.

We don't enforce and only replace here, but better to have the correct contracts in test cases, or add comments regarding contracts that are intentionally incorrect.

File: src/goto-instrument/code_contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: old -> __CPROVER_old

Comment (SaswatPadhi) [reply src/goto-instrument/code_contracts.cpp]: operands -> operand

File: regression/contracts/history-pointer-replace-03/test.desc
Comment (SaswatPadhi) [inline regression/contracts/history-pointer-replace-03/test.desc]: Ideally, this should be a compilation error -- `__CPROVER_old` should not be allowed outside of `__CPROVER_ensures`. This could be done during type checking of `__CPROVER_requires`.

If not, could we generate a helpful error message on verification failure? Also, please add a regex for the expected error message above.

Comment (SaswatPadhi) [reply regression/contracts/history-pointer-replace-03/test.desc]: Missing newline at the end of the file.

File: src/goto-instrument/code_contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp:255]: `auto &` -> `const auto &`

File: src/goto-instrument/code_contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: See my comment above on `std::pair<goto_programt, goto_programt> ensures_pair;`

Then I would replace this entire loop with `check.destructive_append(ensures_pair.second);`

File: regression/contracts/history-pointer-enforce-03/main.c
Comment (SaswatPadhi) [reply regression/contracts/history-pointer-enforce-03/main.c]: Also, `#include <limits.h>` and use `INT_MAX` instead. I think this might fix the Windows build issues?

File: regression/contracts/history-pointer-replace-01/main.c
Comment (SaswatPadhi) [inline regression/contracts/history-pointer-replace-01/main.c]: `#include <limits.h>` and use `INT_MAX` instead. I think this might fix the Windows build issues?

File: regression/contracts/history-pointer-enforce-07/test.desc
Comment (SaswatPadhi) [inline regression/contracts/history-pointer-enforce-07/test.desc]: Looks like exit code is 64 on Windows. So use `^EXIT=(1|64)$`

File: src/goto-instrument/code_contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: See my comment above on `std::pair<goto_programt, goto_programt> ensures_pair`

Then I would replace this entire loop and the next 2 lines with:
```c
    auto lines_to_iterate = ensures_pair.first.instructions.size() + ensures_pair.second.instructions.size();
    goto_program.insert_before_swap(target, ensures_pair.second);
    goto_program.insert_before_swap(target, ensures_pair.first);
```

File: src/goto-instrument/code_contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: "... in the ~~above instruction~~ expression"?

File: src/goto-instrument/code_contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: Can this be `const codet &expression`?

File: src/goto-instrument/code_contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: `auto it`

File: src/goto-instrument/code_contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: Why not store the `dereference_exprt` in the map (after declaring it) in the first place, so we don't have to construct the `dereference_exprt` every time?

File: src/goto-instrument/code_contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: The last argument could be `goto_programt &history`.

File: src/goto-instrument/code_contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: This can be `history.add(goto_programt::make_decl(...))`.

File: src/goto-instrument/code_contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: Similarly `history.add(goto_programt::make_assignment(...))`.

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
The PR overall looks good. I list several comments regarding simplifying the code and clarifying the regression tests.

One meta comment about regression tests: you only check for exit code and verification success and failure. Could you please also add some regexes in the `test.desc` files to check if the contracts are actually being processed? They must be generating some additional assertions, for which we should have regexes.

File: regression/contracts/history-pointer-enforce-03/test.desc
Comment (ArenBabikian) [reply regression/contracts/history-pointer-enforce-03/test.desc]: Oh, this is a mistake. I will fix it.

File: regression/contracts/history-pointer-enforce-07/test.desc
Comment (ArenBabikian) [reply regression/contracts/history-pointer-enforce-07/test.desc]: Thanks!

File: src/goto-instrument/code_contracts.cpp
Comment (ArenBabikian) [reply src/goto-instrument/code_contracts.cpp:255]: I cannot do this because there is a path where I am modifying `op` within the loop.
I might be incorrect.

Comment (SaswatPadhi) [reply src/goto-instrument/code_contracts.cpp:255]: Ah yes, you're doing `op = *op_result;` in the loop.

File: src/goto-instrument/code_contracts.cpp
Comment (ArenBabikian) [reply src/goto-instrument/code_contracts.cpp]: I was able to change it to `codet &expression`. However I did make some slight modifications inside the function and at the call sites. But overally, I believeI was able to provide a certain improvement.

Comment (SaswatPadhi) [reply src/goto-instrument/code_contracts.cpp]: Nice.

I can review the change when after you push it. I'm curious why `const codet &` wouldn't work:
- if you're modifying `expression` within this function, then `codet expression` would have been incorrect in the first place, because that was a copy
- if you're not modifying `expression` within this function, then `const codet &` should work fine

File: regression/contracts/history-pointer-enforce-03/main.c
Comment (ArenBabikian) [reply regression/contracts/history-pointer-enforce-03/main.c]: I have fixed the `INT_MAX` issue. However, regarding the placement of `__CPROVER_ensures`, the current location is what clang is proposing. I will try what you suggested, but it might not work.

Comment (SaswatPadhi) [reply regression/contracts/history-pointer-enforce-03/main.c]: Oh I see. Yeah `clang-format` goes crazy on contract annotations sometimes.

I usually surround contracts with `/* clang-format off */` and `/* clang-format on */`.

File: regression/contracts/history-pointer-replace-03/test.desc
Comment (ArenBabikian) [reply regression/contracts/history-pointer-replace-03/test.desc]: In my upcoming commit, there will be a compilation error if there is a `__CPROVER_old` expressions in a `__CPROVER_requires` clause.

File: regression/contracts/history-pointer-enforce-03/main.c
Comment (ArenBabikian) [reply regression/contracts/history-pointer-enforce-03/main.c]: This positioning of `_CPROVER_ensures` is what clang is proposing. I will try to move it ot the next line, but I might be unable to do this.

File: src/ansi-c/c_typecheck_base.cpp
Comment (SaswatPadhi) [inline src/ansi-c/c_typecheck_base.cpp]: The function call right below this line is `typecheck_expr(requires);`, so this is confusing.

Could you please rename the new function to `disallow_history_variables(requires)`? That would resolve the confusion and the name of your function would also match exactly what it's doing -- which isn't type checking.

File: src/goto-instrument/code_contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: Could you make this `exprt &expr` to avoid a copy. I see that it can't be `const exprt&` since you're modifying it.

Also, this pattern looks bad:
- you pass `expression` to this function, which could be modified within the function
- then you return an `optional<exprt>` from this function
- you replace `expression` if the `optional<exprt>` isn't nil

Why not just pass `exprt &expr` as first argument here and return `void`?
This call that you do later:

```c
  // Find and replace "old" expression in the "expression" variable
  const auto expr_w_replacement = replace_old_parameter(
    expression, parameter2history, location, function, mode, history);
  if(expr_w_replacement.has_value())
  {
    expression = to_code(*expr_w_replacement);
  }
```

could just be one line: `replace_old_parameter(expression, parameter2history, location, function, mode, history);` -- `expression` would be passed by reference, so would be modified within `replace_old_parameter` as needed.

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
The regexes you added to `test.desc` files look good and now we're checking that contracts are actually working as expected.

I have only 2 comments:

File: src/goto-instrument/code_contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: Unused, I think

File: src/util/pointer_expr.h
Comment (SaswatPadhi) [inline src/util/pointer_expr.h]: a predicate -> an expression

`old_exprt` is not a predicate.

File: src/util/pointer_expr.h
Comment (SaswatPadhi) [inline src/util/pointer_expr.h]: a variable -> an expression

The argument can be a deref, for example.

File: src/util/pointer_expr.h
Comment (SaswatPadhi) [inline src/util/pointer_expr.h]: May be rename this to `parameter()` since it need not be just a variable? Can be a deref, for example.

File: src/util/pointer_expr.h
Comment (SaswatPadhi) [inline src/util/pointer_expr.h]: I think this class shouldn't be in `src/util/pointer_expr.h`. Perhaps we should make a new file in this directory: `ghost_variables.h` that would contain all of the `exprt`s for history variables and other ghost variables.

What do you think @feliperodri?

Comment (feliperodri) [reply src/util/pointer_expr.h]: Agreed.

**Review by SaswatPadhi — APPROVED** [review]
LGTM! Thanks for all the changes.

File: src/goto-instrument/code_contracts.cpp
Comment (feliperodri) [inline src/goto-instrument/code_contracts.cpp]: This should go below `<algorithm>`.

**Review by feliperodri — APPROVED** [review]
LGTM.

**Comment by feliperodri** [comment]
Hi @kroening @tautschnig could you review this PR? It's missing code owner approval.

**Comment by kroening** [comment]
While it's conceivable that other languages may eventually offer a similar mechanism, may I suggest to add `old_exprt` to `ansi-c/c_expr.h` for now.

**Comment by ArenBabikian** [comment]
@kroening Thanks for the comment, I have made the required changes.

File: src/ansi-c/c_expr.h
Comment (kroening) [inline src/ansi-c/c_expr.h]: I would fix the ID of `old_exprt` to `ID_old` -- I don't see an extension coming whereby you need several different kinds of "old expressions".

Comment (kroening) [reply src/ansi-c/c_expr.h]: Likewise, this isn't a "base class".

File: src/ansi-c/c_expr.h
Comment (kroening) [inline src/ansi-c/c_expr.h:213]: "Parameter" is wrong terminology here. Given that you don't support symbols yet, the expression that's returned is actually never a parameter. Call it "expression".

File: src/ansi-c/c_typecheck_base.h
Comment (kroening) [inline src/ansi-c/c_typecheck_base.h]: This needs to be
`void disallow_history_variables(const exprt &) const;`
No need for this to be virtual, the parameter must be a const-ref, and the method itself is const.

File: src/ansi-c/c_typecheck_expr.cpp
Comment (kroening) [inline src/ansi-c/c_typecheck_expr.cpp:2630]: operands -> operand

File: src/ansi-c/c_typecheck_base.cpp
Comment (kroening) [inline src/ansi-c/c_typecheck_base.cpp:731]: How about doing that call to `disallow_history_variables` after the typechecking, which would simplify it significantly.

File: src/ansi-c/library/cprover.h
Comment (kroening) [inline src/ansi-c/library/cprover.h:40]: I would make that
`void __CPROVER_old();`
since you want to support arbitrary expressions eventually.

File: src/goto-instrument/code_contracts.cpp
Comment (kroening) [inline src/goto-instrument/code_contracts.cpp]: `PRECONDITION` isn't meant for restrictions on user-input -- the idea is that these are provable.
Throw an exception whenever user input doesn't satisfy the restrictions.

**Review by kroening — APPROVED** [review]

**Review by chrisr-diffblue — APPROVED** [review]

**Pull request merged**
